### DISCUSSION TITLE
Using v5 API to subscribe to context changes

### DISCUSSION THREAD

**Discussion by ts-candide** [statelyai/xstate#4470] [Q&A] [answered]
With v4, `service.onChange((context, previousContext) => { /* ... */ })` can be used to listen for changes to context.

v5 has removed `service.onChange`. How would this behaviour be replicated with `service.subscribe`, assuming I'm using React?

Here's a very basic example of `onChange` being used: https://codesandbox.io/s/hardcore-wind-y7725s?file=/src/index.js

``` jsx
import "./styles.css";
import * as React from "react";
import * as ReactDOM from "react-dom";
import { createMachine, assign } from "xstate";
import { useMachine } from "@xstate/react";

const toggleMachine = createMachine({
  id: "toggle",
  initial: "inactive",
  context: {
    count: 0
  },
  states: {
    inactive: {
      on: { TOGGLE: "active" }
    },
    active: {
      entry: assign({ count: (ctx) => ctx.count + 1 }),
      on: { TOGGLE: "inactive" }
    }
  }
});

function App() {
  const [state, send, service] = useMachine(toggleMachine);
  const active = state.matches("active");
  const { count } = state.context;

  const [logs, setLogs] = React.useState([]);

  React.useEffect(() => {
    const listener = (context, prevContext) => {
      if (context.count !== prevContext.count) {
        setLogs((l) => [
          `Toggle changed from ${prevContext.count} to ${context.count}`,
          ...l
        ]);
      }
    };

    service.onChange(listener);

    return () => {
      service.off(listener);
    };
  }, [service]);

  return (
    <div className="App">
      <h1>XState React Template</h1>
      <h2>Fork this template!</h2>
      <button onClick={() => send("TOGGLE")}>
        Click me ({active ? "✅" : "❌"})
      </button>{" "}
      <code>
        Toggled <strong>{count}</strong> times
      </code>
      <section>
        {logs.map((log, i) => (
          <p key={i}>{log}</p>
        ))}
      </section>
    </div>
  );
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```

**Comment by ts-candide** [comment]
To expand a little on my use case. I am implementing a Shopping Cart, and want to be notified any time an item is added/edited/removed from the basket. The FSM's actions (`addItem` / `editItem` / `removeItem`) are modifying the context in an immutable fashion, meaning that any changes to the context property is a simple equality test (i.e. `context.cartItems !== previousContext.cartItems`). I'm aware that I could call a `cartChanged` action any time the cart is modified in the machine, but this is prone to error, and so I'd rather subscribe to changes in one place.

Perhaps there is a better way to achieve my goal?

**Comment by davidkpiano** [comment, marked as answer]
We've consolidated the API, so you can subscribe to the actor instead and keep a reference to the previous context:

```ts
useEffect(() => {
  let prevContext;

  const sub = actorRef.subscribe(s => {
    console.log(s.context, prevContext);
    prevContext = s.context;
  });

  return sub.unsubscribe;
}, [actorRef]);
```

**Reply by ts-candide** [reply]
Perfect - thank you for your help!

Just for my reference, in v4 the third item returned from `useMachine` is known as a `service` - this is now an `actor` in v5?

**Reply by Andarist** [reply]
In our test cases, we usually call it an `actorRef` nowadays. But calling it an `actor` is, IMHO, fine as well - for practical purposes it doesn't quite matter (just like a distinction between a parameter and an argument usually doesn't matter ;p)
